
**Diplomatura en Ciencia de Datos, Aprendizaje Automático y sus Aplicaciones**

**Edición 2021**

---

# Trabajo práctico entregable - Parte 2

In [18]:
import io
import matplotlib
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn
import statistics 
import math
import scipy
import scipy.stats as stats

seaborn.set_context('talk')

## Lectura del dataset

En la notebook 00 se explican los detalles de la siguiente sección.

In [19]:
url = 'https://cs.famaf.unc.edu.ar/~mteruel/datasets/diplodatos/sysarmy_survey_2020_processed.csv'
df = pd.read_csv(url)

In [20]:
df[:3]

profile_gender  profile_age work_country                    work_province  \
0          Mujer           26    Argentina  Ciudad Autónoma de Buenos Aires   
1         Hombre           29    Argentina                       Corrientes   
2          Mujer           22    Argentina  Ciudad Autónoma de Buenos Aires   

   profile_years_experience  work_years_in_company  \
0                       3.0                    3.0   
1                       5.0                    2.0   
2                       2.0                    0.0   

   work_years_in_current_position  work_people_in_charge_of  \
0                             3.0                         0   
1                             2.0                         4   
2                             0.0                         0   

  profile_studies_level profile_studies_level_state  ...  \
0         Universitario                    En curso  ...   
1         Universitario                    En curso  ...   
2            Secundario                  Completado  ...   

  work_has_violence_situations profile_sexual_orientation  \
0         En mi trabajo actual                 Homosexual   
1                        Jamás               Heterosexual   
2       En un trabajo anterior           Bisexual o queer   

  profile_has_disabilities profile_has_disabilities_hiring_difficulties  \
0                      NaN                                          NaN   
1                   Visual                                           No   
2                      NaN                                           No   

  company_employee_number                          company_main_activity  \
0                501-1000  Servicios / Consultoría de Software / Digital   
1                 201-500                               Otras industrias   
2               2001-5000                               Otras industrias   

  company_recommended company_diversity_policies  \
0                   7                          2   
1                   8                          9   
2                   6                          9   

                              company_extra_benefits  \
0  Capacitaciones y/o cursos, Comidas pagas / sub...   
1  Horarios flexibles, Stock options / RSUs, Viát...   
2  Clases de gimnasia online, Comidas pagas / sub...   

  company_best_companies_city  
0                         NaN  
1                         NaN  
2                         NaN  

[3 rows x 48 columns]

In [21]:
df[['profile_gender', 'salary_monthly_NETO']].groupby('profile_gender').describe()

salary_monthly_NETO                                     \
                             count          mean            std   min   
profile_gender                                                          
Hombre                      4944.0  98836.063558  102159.664237   1.0   
Mujer                        908.0  73501.414465   56256.880181   2.0   
Otros                         31.0  95441.354839   88635.674358  40.0   

                                                       
                    25%      50%       75%        max  
profile_gender                                         
Hombre          53285.0  80000.0  110000.0  2080000.0  
Mujer           43818.0  65000.0   92000.0   800000.0  
Otros           39500.0  74000.0  117500.0   380000.0

In [22]:
df[df.salary_monthly_NETO > 1000]\
  [['profile_gender', 'salary_monthly_NETO']].groupby('profile_gender').describe()

salary_monthly_NETO                                         \
                             count           mean            std      min   
profile_gender                                                              
Hombre                      4815.0  101481.624343  102215.304684   1100.0   
Mujer                        863.0   77328.913778   55082.760570   2000.0   
Otros                         28.0  105660.714286   87235.710235  34000.0   

                                                       
                    25%      50%       75%        max  
profile_gender                                         
Hombre          55000.0  80000.0  110000.0  2080000.0  
Mujer           47850.0  67200.0   93000.0   800000.0  
Otros           43750.0  78900.0  122750.0   380000.0

In [28]:
alpha = 0.05

In [35]:
is_man = df.profile_gender == 'Hombre'

groupA = df[(df.salary_monthly_NETO > 1000) & is_man].salary_monthly_NETO
groupB = df[(df.salary_monthly_NETO > 1000) & ~is_man].salary_monthly_NETO


4815

## Ejercicio 1: Estimación

**Consigna:**  Calcular una estimación puntual y un intervalo de confianza de nivel (1-alpha) para la resta entre la media del salario Neto para Hombres menos la media del salario Neto para otros géneros(diferencia de las medias entre el grupoA y grupoB). 
¿Cómo se relaciona este intervalo de confianza con el test de hipótesis?

In [33]:
def calcEst(groupA, groupB, alpha):
  meanA = groupA.mean()
  meanB = groupB.mean()
  desv_standA = statistics.stdev(groupA)
  desv_standB = statistics.stdev(groupB)
  n_A = len(groupA)
  n_B = len(groupB)
  dif_means = meanA - meanB
  varA = groupA.var()
  varB = groupB.var()
  dif_desv_nA = varA / n_A
  dif_desv_nB = varB / n_B
  raiz_ABsum = math.sqrt(dif_desv_nA + dif_desv_nB)
  interv_conf_sum = dif_means + (stats.norm.ppf((1-alpha/2)) * raiz_ABsum)
  interv_conf_rest = dif_means - (stats.norm.ppf((1-alpha/2)) * raiz_ABsum)
  print('Diferencia entre la media del grupo A y B: ', dif_means)
  print('El intervalo de confianza va entre: ' ,interv_conf_rest," - ",interv_conf_sum)
  print('El estadístico es: ', dif_means/raiz_ABsum)

calcEst(groupA, groupB, alpha)

Diferencia entre la media del grupo A y B:  23262.37339991823
El intervalo de confianza va entre:  18560.667138346143  -  27964.079661490316
El estadístico es:  9.697205976350695


**CONCLUSIÓN**

A partir de los resultados obtenidos anteriormente podemos concluir que la diferencia de las medias entre los salarios es de 23262,37 (pesos) a favor del salario de los hombres. El intervalo de confianza va desde los 18560 hasta los 27964 pesos con un 95% de confianza.

Si el intervalo de confianza no contiene el valor de la hipótesis nula, los resultados son estadísticamente significativos. Como el estadistico esta por fuera del intervalo se puede rechazar la hipotesis nula con una confianza del 95%.


## Ejercicio 2: Test de hipótesis



### 2.1 Formalización

Describir formalmente los distintos compenentes de un test de hipótesis para comprobar si la distribución de los salarios es distinta entre los grupos A y B.



#### Hipotesis nula e hipótesis alternativa
$H_0 $  La media del salario neto de hombres = la media del salario neto de mujeres

$H_1$ La media del salario neto de hombres > la media del salario neto de mujeres

#### Estadístico obtenido en el ejecicio anterior
Estadistico= $$ Z = \frac {xA-xB}{\sqrt{\frac{varA}{nA}+\frac{varB}{nB}}}$$

Esto es asumiendo la hipotesis nula en donde las medias poblacionales son iguales por ende se anulan.

El estadístico es:  9.697205976350695

Dicho estadistico tiene distribucion normal con media 0 y desvio 1.

#### CONCLUSIÓN
Como el estadistico esta por fuera del intervalo se puede rechazar la hipotesis nula con una confianza del 95%  afirmando que hay una diferencia entre la media salarial de ambos grupos a favor del grupo A.

### 2.2 P-valor

1. Calcule el p-valor y decida si rechazar o no la hipótesis nula.
2. Interprete el resultado.
3. Los dos grupos de nuestra muestra tienen tamaños muy distintos. ¿Esto afecta al tests?


In [52]:
scipy.stats.ttest_ind(groupA, groupB, equal_var=False)

Ttest_indResult(statistic=9.697205976350673, pvalue=8.605197322133209e-22)

In [54]:
100-8.605197322133209e-20

100.0

El p-valor nos dice que tan extremo es el valor del estadístico obtenido de la muestra. Nos permite saber que tan estrictos podemos llegar a ser al momento deplantear el mismo contraste de hipótesis pero con distintos niveles de "confianza". 

Si queremos fallar en contra de la hipótesis nula con un 95% de certeza, se debe
cumplir que p =< 0.05
Esta condición se satisface en nuestro ejemplo. El estadístico asociado reportó un valor de lo que aporta evidencia en contra de la hipótesis nula, y nuestro pvalue = 8.605197322133209e-22  al ser menor que alpha decimos que la hipotesis nula es muy poco probable que suceda siendo esta rechazada. 

Si el n fuera mayor el intervalo de confianza se reduce asumiendo el mismo intervalo de confianza, y si este fuese menor el intervalo aumentaria. Para este ejemplo se debe tener un n mayor a 30 en cada grupo lo cual se cumple. 

### [Opcional] 2.3 Potencia del test

Nuestra muestra, ¿era lo suficientemente grande para detectar si existe o no una diferencia entre los grupos?

1. Utilice la función `tt_ind_solve_power` para calcular el tamaño necesario de la muestra para un poder estadístico de 0.8, 0.9 y 0.95, asumiendo una significancia estadística de 0.05.
2. ¿Cómo intepretan el poder estadístico de un test? Dado su conocimiento de dominio sobre los datos, ¿les parece que esta muestra es lo suficientemente grande para ser representativo de la tendencia general? ¿y para utilizarlo en un juicio penal contra una empresa XX por una causa de discriminación?

[Documentación](https://www.statsmodels.org/stable/generated/statsmodels.stats.power.tt_ind_solve_power.html)

NOTA: este análisis debería hacerse ANTES de recolectar los datos.

In [55]:
from statsmodels.stats.power import tt_ind_solve_power

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [56]:
effect_size = (groupA.mean() - groupB.mean()) / groupB.std()
# nobs1=None  - What we want to know
alpha = 0.05
ratio = len(groupB) / len(groupA)

In [57]:
tt_ind_solve_power(effect_size=effect_size, alpha=alpha, power=power, ratio=ratio)

NameError: ignored

## Ejercicio 3: Comunicación y visualización

**Consigna:** Seleccionen un resultado que les parezca relevante a partir de alguno de los ejercicios del entregable. Diseñe e implemente una comunicación en base a este mensaje, en un archivo PDF.

Elija las palabras y visualización más adecuada para que la comunicación sea entendible, efectiva y se dapte a UNA de las siguientes situaciones:

1. Una sección en un artículo de difusión a presentar como parte de una organización sin fines de lucro.
No más de 1 página A4 (o dos si los gráficos son muy grandes).
  1. Ejemplo: Alguna de las secciones [Los ecosistemas de emprendimiento de América Latina y el Caribe frente al COVID-19: Impactos, necesidades y recomendaciones](https://publications.iadb.org/es/los-ecosistemas-de-emprendimiento-de-america-latina-y-el-caribe-frente-al-covid-19-impactos-necesidades-y-recomendaciones), por ejemplo la sección *2.2. Reacciones de los emprendedores*.
  2. Ejemplo: Alguna de las secciones de [The state of gender pay gap in 2021](https://www.payscale.com/data/gender-pay-gap?tk=carousel-ps-rc-job)
  3. Puntos clave: 
    1. Simpleza de los gráficos.
    2. Comunicación en lenguaje simple a personas que no son necesariamente expertos de dominio.
    2. Selección de UNA oración sobre la que se hace énfasis.
    3. No es necesario que mencionen objetivos ni descripciones del conjunto de datos, se supone que eso ya estaría explicado en otras secciones del informe.

2. Una publicación científica o reporte técnico interno. No más de una página A4:
  2. Ejemplo: La sección de resultados de [IZA DP No. 12914: The Impact of a Minimum Wage Change on the Distribution of Wages and Household Income](https://www.iza.org/publications/dp/12914/the-impact-of-a-minimum-wage-change-on-the-distribution-of-wages-and-household-income).
  2. Ejemplo: Alguna de las secciones de [Temporary reduction in daily global CO2 emissions during the COVID-19 forced confinement](https://www.nature.com/articles/s41558-020-0797-x)
  3. Puntos clave:
    3. Nivel de detalle técnico requerido. Es necesario justificar la validez del análisis.
    4. La idea presentada puede ser más compleja. Pueden asumir que la audiencia tiene conocimiento técnico y va a analizar las visualizaciones en detalle. 
    5. Pueden presentar más en detalle las limitaciones del análisis (significancia estadística, etc.)
    2. No es necesario que mencionen objetivos ni descripciones del conjunto de datos, se supone que eso ya estaría explicado en otras secciones del informe.

3. Un tweet (o post de LinkedIn) para la cuenta de su empresa consultora que hace análisis de datos. El objetivo es promocionar un análisis de datos abiertos que van a incluir en su portfolio:
  1. Ejemplo: [Comparación vacunas covid](https://twitter.com/infobeautiful/status/1381577746527236098?s=20)
  2. Ejemplo: [Tweet del BID](https://twitter.com/el_BID/status/1388508583944507396?s=20). Lo valioso de este tweet es que usaron un único número para transmitir un mensaje. Puede ser algo así, o con un gráfico muy simple.
  3. Ejemplo: [Cambio climático](https://twitter.com/UNFCCC/status/1387732156190011394?s=20) Es un muy buen ejemplo, excepto que el gráfico no se lee nada y hay que entrar a la publicación original.
  3. Ejemplo: [¿Cuánto están los programadores en las empresas?](https://www.linkedin.com/posts/denis-rothman-0b034043_tech-career-work-activity-6793861923269054464-gS6y) (No verificamos la veracidad o seriedad de la fuente).
  4. Puntos clave:
    1. Su audiencia no va a mirar la visualización por más de unos segundos, y no tiene conocimiento técnico.
  3. Tienen que incluir además una *breve* descripción de cómo obtuvieron los datos que están presentando, que no entraría en el tweet.
